interfaces/builtin/greengrass_support.go

```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/proc/cgroups r,
+owner @{PROC}/[0-9]*/cgroup r,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+# specific rule for cpuset files
+owner /old_rootfs/sys/fs/cgroup/cpuset/{,system.slice/,user.slice/,user.slice/system.slice/}cpuset.{cpus,mems} rw,
+# TODO: ideally we could include "net_cls,net_prio" inside the one above this that handles
+# blkio, cpuset, memory, etc. but it's not currently clear how to escape a comma inside the pattern...
+owner /old_rootfs/sys/fs/cgroup/net_cls,net_prio/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/net_cls,net_prio/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+# same comment as for "net_cls,net_prio" above but for "cpu,cpuacct" ...
+owner /old_rootfs/sys/fs/cgroup/cpu,cpuacct/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/cpu,cpuacct/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
```

Comment:
These accesses only appear when running the greengrass snap on Ubuntu classic (i.e. specifically Ubuntu 18.04 classic), and also as a command rather than a service. 

I originally only had the `greengrassd` daemon run as an app in the snap, because to run it needs to have certificates installed by the user at runtime. If the certificates weren't available when `greengrassd` runs, it would fail, which means that if `greengrassd` was a proper service in the snap, the snap couldn't be installed as the service would fail to start during install and the install would be aborted. 

I debated removing these now that the necessary functionality to disable services from the install hook landed in snapd 2.36 (see PR https://github.com/snapcore/snapd/pull/5979), and we have the `greengrassd` daemon as a proper service in the snap, but I figured if some day the greengrass snap is adjusted to be a command the user invokes manually again then we may want these accesses.

All that being said, yes Greengrass does appear to just try to read all cgroups (and it does fail if any of the reads are denied), so if you'd prefer we could just use a simpler rule instead.

Reply:
I suggest having policy that works when running greengrassd as a service since that is how it will be delivered. I suspect it will clean up a few things in the policy. The older snap had instructions for setting up the certificates. You could disable the services like you suggest or simply check if the certificates are present and if not log that the service wasn't started because the certs weren't present. Up to you on how to implement the snap, but this seems like a clean way forward policy-wise.

Reply:
I have adjusted this all to only include write accesses to cgroup paths when running as a systemd service.

Reply:
(also I added a rule for read access anywhere and only have write specific paths for the ones it seems to use)